core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -521,18 +513,69 @@ void initializeInternal(URI uri, org.apache.hadoop.conf.Configuration conf) thro
    *
-   * @param uri the uri
-   * @param conf the hadoop conf
+   * @param uri a Alluxio Uri
+   * @param conf a hadoop conf
    * @return whether the details match
    */
-  private boolean connectDetailsMatch(URI uri, org.apache.hadoop.conf.Configuration conf) {
-    // Create the master inquire client that we would have after merging the hadoop conf into
-    // Alluxio Configuration.
+  private boolean connectDetailsMatch(AlluxioURI uri, org.apache.hadoop.conf.Configuration conf) {
     AlluxioConfiguration alluxioConf = new InstancedConfiguration(Configuration.global());
+    // Merge hadoop configuration into Alluxio configuration
     HadoopConfigurationUtils.mergeHadoopConfiguration(conf, alluxioConf);
+
+    // Connection configuration in the Alluxio URI has the highest priority
+    Map<String, Object> alluxioConfProperties = new HashMap<>();
+    if (uri.getAuthorityType() == AlluxioURI.AuthorityType.ZOOKEEPER) {
+      ZookeeperAuthority authority = (ZookeeperAuthority) uri.getAuthority();
+      alluxioConfProperties.put(PropertyKey.ZOOKEEPER_ENABLED.getName(), true);
+      alluxioConfProperties.put(PropertyKey.ZOOKEEPER_ADDRESS.getName(),
+          authority.getZookeeperAddress());
+    } else if (uri.getHost() != null) {
+      alluxioConfProperties.put(PropertyKey.MASTER_HOSTNAME.getName(), uri.getHost());
+      alluxioConfProperties.put(PropertyKey.MASTER_RPC_PORT.getName(), uri.getPort());
+      alluxioConfProperties.put(PropertyKey.ZOOKEEPER_ENABLED.getName(), false);
+      alluxioConfProperties.put(PropertyKey.ZOOKEEPER_ADDRESS.getName(), null);
+    }
+    // Merge connection details in URI into Alluxio configuration
+    alluxioConf.merge(alluxioConfProperties, Source.RUNTIME);
+
     ConnectDetails newDetails = Factory.getConnectDetails(alluxioConf);
+    ConnectDetails oldDetails = FileSystemContext.get()
+        .getMasterInquireClient().getConnectDetails();
+
+    boolean isMatch = newDetails.equals(oldDetails);
+    if (!isMatch) {
+      LOG.warn(ExceptionMessage.DIFFERENT_CONNECTION_DETAILS
+          .getMessage(newDetails, oldDetails));
+    }
 
-    return newDetails.equals(FileSystemContext.get().getMasterInquireClient().getConnectDetails());
+    return isMatch;
+  }
+
+  /**
+   * Sets Alluxio configuration from a uri + hadoop conf.
+   *
+   * @param uri a Alluxio Uri that may contain connection configuration
+   * @param conf the hadoop configuration to merge
+   */
+  private void setAlluxioConfiguration(AlluxioURI uri, org.apache.hadoop.conf.Configuration conf) {
+    // Load Alluxio configuration if any and merge to the one in Alluxio file system. These
+    // modifications to ClientContext are global, affecting all Alluxio clients in this JVM.
+    // We assume here that all clients use the same configuration.
+    HadoopConfigurationUtils.mergeHadoopConfiguration(conf, Configuration.global());
+
+    // Connection configuration in the Alluxio URI has the highest priority
```

Comment:
this method looks a lot like `connectDetailsMatch`. Could we combine them like this?
1) Compute the new Alluxio configuration `newConf`
2) Check whether the connect details match between `newConf` and the global conf
3) If they don't match, merge `newConf` into the global conf

Reply:
This way is clean in code and easier to understand.
but then all the configuration will have the source == RUNTIME, would it be confusing sometimes?

Reply:
Good idea! avoid duplicate computation.
Now I first record the configuration properties map of the input URI.
And pass it to connect details match and initialize internal.